scripts/internal/check_broken_links.py

```diff
@@ -0,0 +1,154 @@
+* Remove duplicates (because regex is not 100% efficient as of now).
+* Check validity of URL, using HEAD request. (HEAD to save bandwidth)
+  Uses requests module for others are painful to use. REFERENCES[9]
+  Handles redirects, http, https, ftp as well.
+
+REFERENCES:
+Using [1] with some modificatons for including ftp
+[1] http://stackoverflow.com/a/6883094/5163807
+[2] http://stackoverflow.com/a/31952097/5163807
+[3] http://daringfireball.net/2010/07/improved_regex_for_matching_urls
+[4] https://mathiasbynens.be/demo/url-regex
+[5] https://github.com/django/django/blob/master/django/core/validators.py
+[6] https://data.iana.org/TLD/tlds-alpha-by-domain.txt
+[7] https://codereview.stackexchange.com/questions/19663/http-url-validating
+[8] https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD
+[9] http://docs.python-requests.org/
+
+"""
+
+from __future__ import print_function
+
+import os
+import re
+import sys
+import time
+
+from concurrent.futures import ThreadPoolExecutor
+import requests
+
+
+HERE = os.path.abspath(os.path.dirname(__file__))
+
+REGEX = r'(?:http|ftp|https)?://' \
+        r'(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+'
+
+# There are some status codes sent by websites on HEAD request.
+# Like 503 by Microsoft, and 401 by Apple
+# They need to be sent GET request
+RETRY_STATUSES = [503, 401, 403]
+
+
+def get_urls(filename):
+    """Extracts all URLs available in specified filename
+    """
+    # fname = os.path.abspath(os.path.join(HERE, filename))
+    # expecting absolute path
+    fname = os.path.abspath(filename)
+    text = ''
+    with open(fname) as f:
+        text = f.read()
+
+    urls = re.findall(REGEX, text)
+    # remove duplicates, list for sets are not iterable
+    urls = list(set(urls))
+    # correct urls which are between < and/or >
+    i = 0
+    while i < len(urls):
+        urls[i] = re.sub("[\*<>\(\)\)]", '', urls[i])
+        i += 1
+
+    return urls
+
+
+def validate_url(url):
+    """Validate the URL by attempting an HTTP connection.
+    Makes an HTTP-HEAD request for each URL.
+    Uses requests module.
+    """
+    try:
+        res = requests.head(url)
+        # some websites deny 503, like Microsoft
+        # and some send 401, like Apple, observations
+        if (not res.ok) and (res.status_code in RETRY_STATUSES):
+            res = requests.get(url)
+        return res.ok
+    except requests.exceptions.RequestException:
+        return False
+
+
+def parallel_validator(urls):
+    """validates all urls in parallel
+    urls: tuple(filename, url)
+    """
+    fails = []  # list of tuples (filename, url)
+    completed = 0
+    total = len(urls)
+    threads = []
+
+    with ThreadPoolExecutor() as executor:
+        for url in urls:
+            fut = executor.submit(validate_url, url[1])
+            threads.append((url, fut))
+
+        # wait for threads to progress a little
+        time.sleep(2)
```

Comment:
I didn't want to call fut.result(), because it blocks the thread, losing the advantage of concurrency.